Add tz string literal

giordano opened the pull request:
This is just a convenient syntax sugar, but I don't know if you planned to use `tz"..."` for something else.

omus commented:
Thanks for the PR. I assume you want this strictly for convenience?

giordano commented:
Yes, it's just for convenience (it saves a few key strokes ;-).  Feel free to reject it if you don't think it's useful, at least it was a few minute exercise for me :-)

omus commented on test/TimeZones.jl:
```diff
@@ -1,5 +1,5 @@
 # Deserialization can cause us to have two immutables that are not using the same memory
-@test TimeZone("Europe/Warsaw") === TimeZone("Europe/Warsaw")
+@test tz"Europe/Warsaw" === "Europe/Warsaw"
```
I expect this will fail. I would just leave the rest as it was.

omus replied:
Maybe add a new test
```julia 
@test tz"Europe/Warsaw" === TimeZone("Europe/Warsaw")
```

giordano replied:
Oops, didn't mean to remove the `TimeZone` on the right side.  Thanks

omus replied:
I'd like to keep the original test too since it's directly testing for what is mentioned in the comment.

omus commented on docs/types.md:
````diff
@@ -3,7 +3,15 @@
+Europe/Warsaw (UTC+1/UTC+2)
+```
+
+You can also create a `TimeZone` by using a `tz` string literal:
````
"using the `tz` string macro"

giordano replied:
I used the terminology adopted here: https://docs.julialang.org/en/latest/manual/metaprogramming.html#Non-Standard-String-Literals-1 but I don't have a strong opinion, will change following your suggestion.

omus commented on src/TimeZones.jl:
```diff
@@ -76,6 +77,22 @@ function TimeZone(str::AbstractString)
+"""
+    @tz_str str -> TimeZone
+
+Constructs the `TimeZone` of `str` with `TimeZone` function, which see.
```
"which see"?

giordano replied:
This was done on purpose
* https://en.wiktionary.org/wiki/q.v.
* https://www.urbandictionary.com/tags.php?tag=which%20see

I saw used often elsewhere, but if you think it's difficult to understand, please suggest me something more common (I'm not a native English speaker).

giordano commented:
I fixed the test (now it passes on my machine; previously I run the tests on a different version of the patch, not the final one), changed "string literal" to "string macro" and rephrased the docstring of the macro.

giordano commented:
Probably it doesn't play nicely with `TimeZones.TZData.compile`?

omus commented on src/TimeZones.jl:
```diff
@@ -76,6 +77,25 @@ function TimeZone(str::AbstractString)
 
+
+"""
+    @tz_str str -> TimeZone
```
After looking up the docstring for `@r_str` it appears this should just be `@tz_str -> TimeZone`

omus commented on src/TimeZones.jl:
````diff
@@ -76,6 +77,25 @@ function TimeZone(str::AbstractString)
+```
+"""
+macro tz_str(str)
+    return :( TimeZone($str) )
````
Could you remove the whitespace padding inside the parenthesis? i.e. Change to `:(TimeZone($str))`

giordano replied:
The docstring of `macro` itself puts spaces inside the quoting parentheses, but I will change this.

omus commented on test/TimeZones.jl:
```diff
@@ -37,13 +37,13 @@ if lowercase(get(ENV, "CI", "false")) == "true"
 
     TimeZones.TZData.compile(max_year=2200)
-    new_warsaw = TimeZone("Europe/Warsaw")
+    new_warsaw = tz"Europe/Warsaw"
```
Lets leave this line and the [above line](https://github.com/JuliaTime/TimeZones.jl/pull/64/files#diff-72390e967fd3f6573a6058eba616722aR40) the way the were.

giordano replied:
Actually this test was fundamental to spot that previous macro definition wouldn't work after a `TimeZones.TZData.compile`.

omus commented:
> Probably it doesn't play nicely with TimeZones.TZData.compile?

Oh, I didn't catch the problem you were running into before. Since macros execute at parse time they typically don't usually take into account runtime state. I think having the `tz_str` macro evaluate the time zone at parse time is probably best since this can help improve performance when using the `tz_str` macro in a loop. This will cause the macro not to work well with the re-compiling the time zone data but I only expect `compile` to be called during: build, interactive sessions, or during tests.

Here's a quick demo that shows the performance gains of evaluating at parse time:

```julia
julia> using TimeZones, BenchmarkTools

julia> macro tza_str(str)
           TimeZone(str)
       end;

julia> macro tzb_str(str)
           :(TimeZone($str))
       end;

julia> tz = TimeZone("UTC");  # Pre-loads the time zone

julia> function demo_a()
           for i = 1:20000
               ZonedDateTime(i, 4, tza"UTC")
           end
       end;

julia> function demo_b()
           for i = 1:20000
               ZonedDateTime(i, 4, tzb"UTC")
           end
       end;

julia> a = @benchmark demo_a()

BenchmarkTools.Trial: 
  samples:          7876
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  625.00 kb
  allocs estimate:  20000
  minimum time:     482.29 μs (0.00% GC)
  median time:      503.79 μs (0.00% GC)
  mean time:        634.22 μs (7.66% GC)
  maximum time:     55.14 ms (0.00% GC)

julia> b = @benchmark demo_b()
BenchmarkTools.Trial: 
  samples:          1102
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  1.22 mb
  allocs estimate:  60000
  minimum time:     3.74 ms (0.00% GC)
  median time:      4.13 ms (0.00% GC)
  mean time:        4.54 ms (2.71% GC)
  maximum time:     13.79 ms (0.00% GC)

julia> judge(median(a), median(b))
BenchmarkTools.TrialJudgement: 
  time:   -87.80% => improvement (5.00% tolerance)
  memory: -50.00% => improvement (1.00% tolerance)
```

giordano commented:
`tzb_str` macro isn't slower than `TimeZone` (it's probably very slightly faster, but nothing really noticeable).  `tza_str` is faster but doesn't allow for changing state.  Personally I'd go for a solution that works in any case (and isn't bad performance-wise), but the decision is up to you ;-)  Just tell me what to do.

omus commented:
Lets go with the `tza_str` version. Once that is changed and the tests are updated I think we're good.

giordano commented:
Ok, macro definition switched back to `macro tz_str(str) TimeZone(str) end`.  Now the test checks that a `TimeZone` constructed with `tz"..."` doesn't change after re-compilation, so you'll be warned if this behavior will change in the future.
